### PR TITLE
feature/SMC-2954

### PR DESCRIPTION
**Pull request by southbite**

### CODE REVIEW

**Review by JensEggers — APPROVED** [review]

File: lib/services/security/service.js [hunk 1]
```diff
@@ -1311,9 +1336,10 @@ function generateSession(user, sessionId, credentials, tokenLogin) {
   this.__profileSession(session); //session ttl, activity threshold and user effective permissions are set here
 
   session.permissionSetKey = this.generatePermissionSetKey(session.user, session);
-
+  // if (additionalInfo) session = {...session, ...additionalInfo}
+  additionalInfo = additionalInfo || {};
   if (tokenLogin) session.token = tokenLogin.token;
-  else session.token = this.generateToken(session);
+  else session.token = this.generateToken({ ...session, ...additionalInfo }); //Don't want additional info to stay around unencrypted (?)
```
[the remark this answers is not in the document]
Comment (JancoHY) [reply lib/services/security/service.js:1348]: done

File: lib/services/security/authentication/provider-base.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  }
+
+  userCredsLogin(credentials, sessionId, callback) {
+    if (this.__checkLockedOut(credentials.username))
+      return this.accessDenied('Account locked out', callback);
+    return this.__providerCredsLogin(credentials, sessionId, callback);
+  }
+
+  __providerCredsLogin(credentials, sessionId, callback) {
+    return this.accessDenied('Authentication Provider not set up correctly.', callback);
```
[the remark this answers is not in the document]
Comment (JancoHY) [reply lib/services/security/authentication/provider-base.js]: done

File: lib/services/security/authentication/provider-base.js [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+      //Anything further is dealt with in the specific provider
+      return this.__providerTokenLogin(credentials, previousSession, sessionId, callback);
+    } catch (e) {
+      return this.__loginFailed(username, 'Invalid credentials', e, callback);
+    }
+  }
+
+  __providerTokenLogin(username, credentials, callback) {
+    return this.accessDenied('Authentication Provider not set up correctly.', callback);
```
[the remark this answers is not in the document]
Comment (JancoHY) [reply lib/services/security/authentication/provider-base.js]: done

File: README.md [hunk 4]
```diff
@@ -1854,6 +1854,75 @@ serviceInstanceLocked.services.session.localClient({
...
+you can give a path relative to ./lib/services/security/authentication, where any native auth-providers will be stored.
+Example: Happn-3 and 2 other providers, otherAuthProvider default
+```javascript
+const serviceConfig = {
+  secure: true,
+  services: {
+    security: {
+      config: {
+        authProviders:{
+          happn3 :'./happn3-provider.js',
```
[the remark this answers is not in the document]
Comment (JancoHY) [reply README.md]: done

File: lib/services/security/authentication/happn3-provider.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
+let AuthProvider = require('./provider-base');
+module.exports = class Happn3AuthProvider extends AuthProvider {
+  constructor(happn, config) {
+    super(happn, config);
+  }
+
+  static create(happn, config) {
+    return new Happn3AuthProvider(happn, config);
```
[the remark this answers is not in the document]
Comment (JancoHY) [reply lib/services/security/authentication/happn3-provider.js]: done

File: lib/services/security/service.js [hunk 6]
```diff
@@ -631,6 +642,20 @@ function __initializeSessionManagement(config) {
...
 
+function __initializeAuthProviders(config) {
+  let authProviders = config.authProviders || { happn3: 'happn3-provider' };
+  let defaultAuthProvider = config.defaultAuthProvider || 'happn3';
+  this.authProviders = {};
+  Object.keys(authProviders).forEach(key => {
+    let provider = authProviders[key];
+    let internalPath = path.resolve(__dirname, `./authentication/${provider}.js`);
+    this.authProviders[key] = fs.existsSync(internalPath)
+      ? require(internalPath).create(this.happn, config)
```
[the remark this answers is not in the document]
Comment (JancoHY) [reply lib/services/security/service.js]: This is being handled in BaseAuthProvider

**Pull request merged**